src/porepy/numerics/fv/upwind.py

```diff
@@ -456,33 +507,66 @@ def assemble_matrix_rhs(
         sd_primary: pp.Grid,
         sd_secondary: pp.Grid,
         intf: pp.MortarGrid,
-        data_primary: Dict,
-        data_secondary: Dict,
-        data_intf,
+        data_primary: dict,
+        data_secondary: dict,
+        data_intf: dict,
         matrix: sps.spmatrix,
-    ) -> Tuple[sps.spmatrix, np.ndarray]:
-        """
-        Construct the matrix (and right-hand side) for the coupling conditions.
-        Note: the right-hand side is not implemented now.
+    ) -> tuple[sps.spmatrix, np.ndarray]:
+        """Construct the matrix (and right-hand side) for the coupling conditions.
+
+        Note:
+            The right-hand side is not implemented now.
+
+        In the following, we denote with "primary" the higher-dimensional domain and
+        with "secondary" the lower-dimensional domain.
+
+        We assume the following two sub-dictionaries to be present in the ``data_intf``
+        dictionary:
+            - parameter_dictionary, storing all parameters. Stored in
+              ``data_intf[pp.PARAMETERS][self.keyword]``.
+            - matrix_dictionary, for storage of discretization matrices. Stored in
+              ``data_intf[pp.DISCRETIZATION_MATRICES][self.keyword]``.
+
+        parameter_dictionary contains the entries:
+            - ``self._flux_array_key``: :class:`~numpy.ndarray` of
+              ``shape=(sd_primary.num_faces,)``. Normal velocity at each face of the
+              primary grid, weighted by the face area.
+
+        matrix_dictionary contains the following entries:
+            - ``self.inv_trace_primary_matrix_key``: :class:`~scipy.sparse.csr_matrix`
+              of ``shape=(sd_primary.num_faces, sd_primary.num_cells)``. Inverse trace
+              operator.
+            - ``self.trace_primary_matrix_key``: :class:`~scipy.sparse.csr_matrix` of
+              ``shape=(sd_primary.num_faces, sd_primary.num_cells)``. Trace operator.
+            - ``self.upwind_primary_matrix_key``: :class:`~scipy.sparse.csr_matrix` of
+                ``shape=(sd.num_faces, sd.num_cells)``. Upwind matrix for the
+                primary domain.
+            - ``self.upwind_secondary_matrix_key``: :class:`~scipy.sparse.csr_matrix`
+                of ``shape=(sd.num_faces, sd.num_faces)``. Upwind matrix for the
+                secondary domain.
+            - ``self.flux_matrix_key``: :class:`~scipy.sparse.csr_matrix`
+                of ``shape=(sd.num_faces, sd.num_faces)``. Flux matrix.
+            - ``self.mortar_discr_matrix_key``: :class:`~scipy.sparse.csr_matrix`
+                of ``shape=(intf.num_cells, intf.num_cells)``. Identity matrix for the
+                mortar variable.
```

Comment:
There is quite a bit of duplicate documentation in this file now. Sometimes, we prefer referring to docstrings of other function. You may want to consider this in the future (not this file, though).

Reply:
Will keep this in mind in the future!